tests/test_density_matrix.py

```diff
@@ -0,0 +1,397 @@
+
+        dm = DensityMatrix(data=data)
+        dm.ptrace((1,))
+        expected_matrix = np.array([[1, 0], [0, 0]])
+        assert np.allclose(dm.rho, expected_matrix)
+
+        dm = DensityMatrix(data=data)
+        dm.ptrace((0, 1))
+        expected_matrix = np.array([1])
+        assert np.allclose(dm.rho, expected_matrix)
+
+        dm = DensityMatrix(nqubit=4)
+        dm.ptrace((0, 1, 2))
+        expected_matrix = np.array([[1, 1], [1, 1]]) / 2
+        assert np.allclose(dm.rho, expected_matrix)
+
+        dm = DensityMatrix(nqubit=4)
+        dm.ptrace((0, 1, 2, 3))
+        expected_matrix = np.array([1])
+        assert np.allclose(dm.rho, expected_matrix)
+
+        psi = np.kron(np.kron(np.array([1, np.sqrt(2)]) / np.sqrt(3), np.array([1, 0])), np.array([0, 1]))
+        data = np.outer(psi, psi)
+        dm = DensityMatrix(data=data)
+        dm.ptrace((2,))
+        expected_matrix = np.array(
+            [[1 / 3, 0, np.sqrt(2) / 3, 0], [0, 0, 0, 0], [np.sqrt(2) / 3, 0, 2 / 3, 0], [0, 0, 0, 0]]
+        )
+        assert np.allclose(dm.rho, expected_matrix)
+
+
+class DensityMatrixBackendTest(unittest.TestCase):
+    """Test for DensityMatrixBackend class."""
+
+    def test_init_fail(self):
+        with self.assertRaises(TypeError):
+            DensityMatrixBackend()
+
+    def test_init_success(self):
+        circ = Circuit(1)
+        circ.rx(0, np.pi / 2)
+        pattern = circ.transpile()
+        backend = DensityMatrixBackend(pattern)
+        assert backend.pattern == pattern
+        assert backend.results == pattern.results
+        assert backend.state is None
+        assert backend.node_index == []
+        assert backend.Nqubit == 0
+        assert backend.max_qubit_num == 12
+
+    def test_add_nodes(self):
+        circ = Circuit(1)
+        pattern = circ.transpile()
+        backend = DensityMatrixBackend(pattern)
+        backend.add_nodes([0, 1])
+        expected_matrix = np.array([0.25] * 16).reshape(4, 4)
+        np.testing.assert_allclose(backend.state.rho, expected_matrix)
+
+    def test_entangle_nodes(self):
+        circ = Circuit(1)
+        pattern = circ.transpile()
+        backend = DensityMatrixBackend(pattern)
+        backend.add_nodes([0, 1])
+        backend.entangle_nodes((0, 1))
+        expected_matrix = np.array([[1, 1, 1, -1], [1, 1, 1, -1], [1, 1, 1, -1], [-1, -1, -1, 1]]) / 4
+        np.testing.assert_allclose(backend.state.rho, expected_matrix)
+
+        backend.entangle_nodes((0, 1))
+        np.testing.assert_allclose(backend.state.rho, np.array([0.25] * 16).reshape(4, 4))
+
+    def test_measure(self):
+        circ = Circuit(1)
+        circ.rx(0, np.pi / 2)
+        pattern = circ.transpile()
+
+        backend = DensityMatrixBackend(pattern)
+        backend.add_nodes([0, 1, 2])
+        backend.entangle_nodes((0, 1))
+        backend.entangle_nodes((1, 2))
+        backend.measure(backend.pattern.seq[-4])
+
+        expected_matrix_1 = np.zeros((4, 4))
```

Comment:
how about np.ones((4, 4))/2. ?

Reply:
If you mean np.ones((2,2))/2, can be refactored as one-liner. Thanks! (in the next commit I'll push this change)
`expected_matrix_1 = np.kron(np.array([[1, 0], [0, 0]]), np.ones((2, 2)) / 2)`

w.r.t this, `expected_matrix_2` is also refactored as one-liner
`expected_matrix_2 = np.kron(np.array([[0, 0], [0, 1]]), np.array([[0.5, -0.5], [-0.5, 0.5]]))`